server/jetstream.go

```diff
@@ -829,32 +854,83 @@ func (a *Account) JetStreamEnabled() bool {
 	return enabled
 }
 
-// Updates accounting on in use memory and storage.
+func (jsa *jsAccount) remoteUpdateUsage(sub *subscription, c *client, subject, _ string, msg []byte) {
+	const usageSize = 16
+
+	jsa.mu.Lock()
+	s := jsa.js.srv
+	if len(msg) != usageSize {
+		jsa.mu.Unlock()
+		s.Warnf("Received remote usage update that is wrong size: %d vs %d", len(msg), usageSize)
+		return
+	}
+	var rnode string
+	if li := strings.LastIndexByte(subject, btsep); li > 0 && li < len(subject) {
+		rnode = subject[li+1:]
+	}
+	if rnode == _EMPTY_ {
+		jsa.mu.Unlock()
+		s.Warnf("Received remote usage update with no remote node")
+		return
+	}
+	var le = binary.LittleEndian
+	memUsed, storeUsed := int64(le.Uint64(msg[0:])), int64(le.Uint64(msg[8:]))
+
+	if jsa.rusage == nil {
+		jsa.rusage = make(map[string]*jsaUsage)
+	}
+	// Update the usage for this remote.
+	if usage := jsa.rusage[rnode]; usage != nil {
+		// Decrement our old values.
+		jsa.memTotal -= usage.mem
+		jsa.storeTotal -= usage.store
+		usage.mem, usage.store = memUsed, storeUsed
+	} else {
+		jsa.rusage[rnode] = &jsaUsage{memUsed, storeUsed}
+	}
+	jsa.memTotal += memUsed
+	jsa.storeTotal += storeUsed
+
+	jsa.mu.Unlock()
+}
+
+// Updates accounting on in use memory and storage. This is called from locally
+// by the lower storage layers.
 func (jsa *jsAccount) updateUsage(storeType StorageType, delta int64) {
-	// TODO(dlc) - atomics? snapshot limits?
 	jsa.mu.Lock()
 	if storeType == MemoryStorage {
-		jsa.memUsed += delta
+		jsa.usage.mem += delta
+		jsa.memTotal += delta
 	} else {
-		jsa.storeUsed += delta
+		jsa.usage.store += delta
+		jsa.storeTotal += delta
+	}
+	// Publish our local updates if in clustered mode.
+	if jsa.js != nil && jsa.js.cluster != nil && jsa.js.srv != nil {
+		s, b := jsa.js.srv, make([]byte, 16)
+		var le = binary.LittleEndian
+		le.PutUint64(b[0:], uint64(jsa.usage.mem))
```

Comment:
I understand that this is called often, hence the desire to make it binary encoded, however, this is pretty limiting future changes (due to the fact that the receiving side would bail out if the size of the thing we send here were different). I wonder if we need every single updates or should do it say every second, in which case using JSON would be more extensible.

Reply:
I thought about that quite a bit. We will only ever send absolutes of usage for the account per server. We could add a version in there but not worth it imo.

Also thought about not sending everyone, and came up with something pretty complicated to speed up as you got close to the limit etc. but abandoned it for the simpler version. 

We could slow it down if need be in the future but for now I prefer the send everytime version.